modules/compiler/spirv-ll/source/builder_debug_info.cpp

```diff
@@ -15,29 +15,2466 @@
+};
+
+template <>
+llvm::Expected<llvm::MDNode *> DebugInfoBuilder::translate(
+    const DebugTypeBasic *op) {
+  std::optional<std::string> name = module.getDebugString(op->Name());
+  if (!name) {
+    return makeStringError(
+        "Could not find OpString 'Name' for DebugTypeBasic " +
+        getIDAsStr(op->IdResult(), &module));
+  }
+
+  auto encoding_iter = DebugEncodingMap.find(op->Encoding());
+  if (encoding_iter == DebugEncodingMap.end() || encoding_iter->second == 0) {
+    return getDIBuilder(op).createUnspecifiedType(*name);
+  }
+  llvm::dwarf::TypeKind encoding = encoding_iter->second;
+
+  auto size_or_error = getConstantIntValue(op->Size());
+  if (auto err = size_or_error.takeError()) {
+    return err;
+  }
+  // Without a size, we can't create a type.
+  if (!size_or_error.get()) {
+    return nullptr;
+  }
+  uint64_t size = *size_or_error.get();
+
+  return getDIBuilder(op).createBasicType(*name, size, encoding);
+}
+
+class DebugTypePointer : public OpExtInst {
+ public:
+  DebugTypePointer(OpCode const &other) : OpExtInst(other) {}
+  spv::Id BaseType() const { return getOpExtInstOperand(0); }
+  uint32_t StorageClass() const {
+    return static_cast<spv::StorageClass>(getOpExtInstOperand(1));
+  }
+  uint32_t Flags() const { return getOpExtInstOperand(2); }
+};
+
+template <>
+llvm::Expected<llvm::MDNode *> DebugInfoBuilder::translate(
+    const DebugTypePointer *op) {
+  auto base_ty = translateDebugInst<llvm::DIType>(op->BaseType());
+  if (auto err = base_ty.takeError()) {
+    return err;
+  }
+  std::optional<unsigned> addrspace;
+  auto addrspace_or_error =
+      module.translateStorageClassToAddrSpace(op->StorageClass());
+  if (auto err = addrspace_or_error.takeError()) {
+    // Silently consume this error. We know that llvm-spirv will use ~0 to
+    // represent "no address space", despite this being invalid SPIR-V.
+    llvm::consumeError(std::move(err));
+  } else {
+    addrspace = addrspace_or_error.get();
+  }
+
+  uint32_t flags = op->Flags();
+  llvm::DIType *type = nullptr;
+  llvm::DIBuilder &dib = getDIBuilder(op);
+
+  if (flags & OpenCLDebugInfo100FlagLValueReference) {
+    type = dib.createReferenceType(llvm::dwarf::DW_TAG_reference_type, *base_ty,
+                                   /*size*/ 0, /*align*/ 0, addrspace);
+  } else if (flags & OpenCLDebugInfo100FlagRValueReference) {
+    type = dib.createReferenceType(llvm::dwarf::DW_TAG_rvalue_reference_type,
+                                   *base_ty, /*size*/ 0,
+                                   /*align*/ 0, addrspace);
+  } else {
+    // This is 32, 64, or 0 if no memory model is specified.
+    uint64_t size = module.getAddressingModel();
+    type = dib.createPointerType(*base_ty, size, /*align*/ 0, addrspace);
+  }
+
+  if (flags & OpenCLDebugInfo100FlagObjectPointer) {
+    type = dib.createObjectPointerType(type);
+  } else if (flags & OpenCLDebugInfo100FlagArtificial) {
+    type = dib.createArtificialType(type);
+  }
```

Comment:
These flags are not documented at all, are they? Given that this specifically calls out *object* pointers, and that DPC++ supports function pointers as an extension, could you add a test for whatever debug info DPC++ generates for those? As long as that is handled sensibly, even if debug information is left out for it, it looks good to me, thank you.

Reply:
Oh, I just realised this maps to DWARF's `DW_AT_object_pointer` which does not mean object pointers in general, it just means the implicit `this` object parameter of non-static member functions. There is otherwise no difference between object and function pointers. Feel free to ignore this, then.